### PR TITLE
test: [POM] Migrate hardware wallet e2e tests to follow Page Object Model

### LINKED ISSUE #28808: [POM] Create base classes for hardware wallet e2e tests
### What is this about?

Create base classes for hardware wallet e2e tests

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [ ] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by chloeYue**
## **Description**

- Create base pages for hardware wallet related pages
- Migrate hardware wallet  e2e tests to Page Object Model
- Remove dead code (We should not keep unused functions in the codebase because they can lead to several issues, such as increased maintenance overhead, added complexity, potential bugs, and misleading information.)

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27155?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/28808

## **Manual testing steps**
Check code readability, make sure tests pass.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: test/e2e/page-objects/pages/test-dapp.ts [hunk 1]
```diff
@@ -657,18 +663,33 @@ class TestDapp {
 
   /**
    * Sign a message with the signTypedDataV4 method.
+   *
+   * @param confirmationRedesign - Indicates whether the redesigned signature confirmation flow is used. Defaults to false.
    */
-  async signTypedDataV4() {
+  async signTypedDataV4(confirmationRedesign: boolean = false) {
```
Comment (chloeYue) [inline test/e2e/page-objects/pages/test-dapp.ts:669]: Keep the `confirmationRedesign` as a parameter as a temporary solution, will do the refacto once the toggle is removed, as planned in [MetaMask-planning issue #303](https://github.com/MetaMask/MetaMask-planning/issues/303).

File: test/e2e/tests/hardware-wallets/lattice-connect.spec.ts [hunk 2]
```diff
@@ -2,8 +2,12 @@ import { strict as assert } from 'assert';
...
-import { withFixtures, unlockWallet } from '../../helpers';
+import { withFixtures } from '../../helpers';
 import { isManifestV3 } from '../../../../shared/modules/mv3.utils';
+import AccountListPage from '../../page-objects/pages/account-list-page';
+import ConnectHardwareWalletPage from '../../page-objects/pages/hardware-wallet/connect-hardware-wallet-page';
+import HeaderNavbar from '../../page-objects/pages/header-navbar';
+import { loginWithBalanceValidation } from '../../page-objects/flows/login.flow';
 
 describe('Lattice hardware wallet @no-mmi', function (this: Suite) {
   it('connects to lattice hardware wallet', async function () {
```
Comment (hjetpoluru) [inline test/e2e/tests/hardware-wallets/lattice-connect.spec.ts:13]: small nit 
```suggestion
  it('lattice page rendering validation', async function () {
```

File: test/e2e/tests/hardware-wallets/trezor-account.spec.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+import { KNOWN_PUBLIC_KEY_ADDRESSES } from '../../../stub/keyring-bridge';
+import AccountListPage from '../../page-objects/pages/account-list-page';
+import ConnectHardwareWalletPage from '../../page-objects/pages/hardware-wallet/connect-hardware-wallet-page';
+import HeaderNavbar from '../../page-objects/pages/header-navbar';
+import HomePage from '../../page-objects/pages/home/homepage';
+import SelectTrezorAccountPage from '../../page-objects/pages/hardware-wallet/select-trezor-account-page';
+import { loginWithBalanceValidation } from '../../page-objects/flows/login.flow';
+
+describe('Trezor Hardware', function () {
+  it('derives the correct accounts', async function () {
```
Comment (hjetpoluru) [inline test/e2e/tests/hardware-wallets/trezor-account.spec.ts:13]: I think this scenario could be combined with the next one.

**Review by hjetpoluru — DISMISSED** [review]
Nice job @chloeYue, I notice minot observation otherwise everything is looking good.

File: test/e2e/tests/hardware-wallets/lattice-connect.spec.ts (hunk 2 above)
Comment (chloeYue) [reply test/e2e/tests/hardware-wallets/lattice-connect.spec.ts:13]: Done, thanks !

File: test/e2e/tests/hardware-wallets/trezor-account.spec.ts (hunk 3 above)
Comment (chloeYue) [reply test/e2e/tests/hardware-wallets/trezor-account.spec.ts]: Make sense, addressed in commit: [bc3c4e1](https://github.com/MetaMask/metamask-extension/pull/28768/commits/bc3c4e1238354120b59aa9f5ead6afb06a6aa629)
Thanks !

**Review by hjetpoluru — DISMISSED** [review]

File: test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import { Driver } from '../../../webdriver/driver';
+
+class ConnectHardwareWalletPage {
```
Comment (dbrans) [inline test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts:3]: Could you add some comments throughout these new files?
E.g.
```suggestion
/**
 * Represents the page for connecting hardware wallets.
 * This page allows users to initiate connections with various hardware wallet types.
 */
class ConnectHardwareWalletPage {
```

File: test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  private readonly connectTrezorButton = '[data-testid="connect-trezor-btn"]';
+
+  private readonly continueButton = { text: 'Continue', tag: 'button' };
+
+  constructor(driver: Driver) {
+    this.driver = driver;
+  }
+
+  async check_pageIsLoaded(): Promise<void> {
```
Comment (dbrans) [inline test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts:21]: Could you clarify the use of the underscore separator and the `check_` prefix? Are we formalizing it as a pattern? If so, would it make sense to document this for contributors?

File: test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    console.log('Connect hardware wallet page is loaded');
+  }
+
+  async openConnectLatticePage(): Promise<void> {
+    console.log(`Open connect lattice page`);
+    await this.driver.clickElement(this.connectLatticeButton);
+    await this.driver.clickElement(this.continueButton);
+  }
+
+  async openConnectTrezorPage(): Promise<void> {
```
Comment (dbrans) [inline test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts:43]:
```suggestion
/**
   * Opens the connect Trezor page and proceeds with the connection flow.
   */
   async openConnectTrezorPage(): Promise<void> {
```

File: test/e2e/page-objects/pages/hardware-wallet/select-trezor-account-page.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
+import { Driver } from '../../../webdriver/driver';
+
+class SelectTrezorAccountPage {
```
Comment (dbrans) [inline test/e2e/page-objects/pages/hardware-wallet/select-trezor-account-page.ts:3]: Same comments here: this file could use more code comments

File: test/e2e/tests/hardware-wallets/trezor-account.spec.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        const accountListPage = new AccountListPage(driver);
+        await accountListPage.check_pageIsLoaded();
+        await accountListPage.openConnectHardwareWalletModal();
+
+        const connectHardwareWalletPage = new ConnectHardwareWalletPage(driver);
+        await connectHardwareWalletPage.check_pageIsLoaded();
+        await connectHardwareWalletPage.openConnectTrezorPage();
+
+        const selectTrezorAccountPage = new SelectTrezorAccountPage(driver);
+        await selectTrezorAccountPage.check_pageIsLoaded();
```
Comment (dbrans) [inline test/e2e/tests/hardware-wallets/trezor-account.spec.ts:35]: Noticed this setup code is repeated and isn’t specific to this test. What do you think about extracting it into a setup method, maybe something like `openSelectTrezorAccountPage`, or even a reusable flow?

File: test/e2e/page-objects/pages/account-list-page.ts [hunk 9]
```diff
@@ -377,6 +385,15 @@ class AccountListPage {
     await this.driver.waitForSelector(this.addEthereumAccountButton);
   }
 
+  async openConnectHardwareWalletModal(): Promise<void> {
```
Comment (dbrans) [inline test/e2e/page-objects/pages/account-list-page.ts:388]: Hey Chloe,

Could we add comments to the methods in this file to make the code easier to follow?

Also, would it make sense for this method to return the `ConnectHardwareWallet` page directly? It seems like that's what every caller ends up creating right after. Curious to hear your thoughts!

File: test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts (hunk 4 above)
Comment (chloeYue) [reply test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts:7]: Done, thanks !

File: test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts (hunk 5 above)
Comment (chloeYue) [reply test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts:25]: yes we formalize it as a pattern and have the naming convention in the contributor documentation.https://github.com/MetaMask/contributor-docs/blob/main/docs/testing/e2e/extension-e2e-guidelines.md#check_-methods

File: test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts (hunk 6 above)
Comment (chloeYue) [reply test/e2e/page-objects/pages/hardware-wallet/connect-hardware-wallet-page.ts]: Done, thanks !

File: test/e2e/tests/hardware-wallets/trezor-account.spec.ts (hunk 8 above)
Comment (chloeYue) [reply test/e2e/tests/hardware-wallets/trezor-account.spec.ts:35]: Thanks for the suggestion Derek. I am currently focusing on creating base pages and methods for hardware wallet pages. For the next steps to better organize the tests, I will leave it to hardware wallet team to decide how to better extract based on their needs.

File: test/e2e/page-objects/pages/hardware-wallet/select-trezor-account-page.ts (hunk 7 above)
Comment (chloeYue) [reply test/e2e/page-objects/pages/hardware-wallet/select-trezor-account-page.ts:7]: Added comment in commit : [5c6702b](https://github.com/MetaMask/metamask-extension/pull/28768/commits/5c6702bea79f3d7ea759daf8e88a5d81c830b1b9)

**Review by dbrans — APPROVED** [review]
Nice work, Chloe!

File: test/e2e/page-objects/pages/home/asset-list.ts [hunk 10]
```diff
@@ -156,30 +147,6 @@ class AssetListPage {
     );
   }
 
-  async selectNetworkFilterAllNetworks(): Promise<void> {
```
Comment (seaona) [inline test/e2e/page-objects/pages/home/asset-list.ts:159]: nice removals!

File: test/e2e/tests/hardware-wallets/lattice-connect.spec.ts [hunk 11]
```diff
@@ -2,39 +2,33 @@ import { strict as assert } from 'assert';
...
+import { withFixtures } from '../../helpers';
 import { isManifestV3 } from '../../../../shared/modules/mv3.utils';
+import AccountListPage from '../../page-objects/pages/account-list-page';
+import ConnectHardwareWalletPage from '../../page-objects/pages/hardware-wallet/connect-hardware-wallet-page';
+import HeaderNavbar from '../../page-objects/pages/header-navbar';
+import { loginWithBalanceValidation } from '../../page-objects/flows/login.flow';
 
 describe('Lattice hardware wallet @no-mmi', function (this: Suite) {
-  it('connects to lattice hardware wallet', async function () {
+  it('lattice page rendering validation', async function () {
```
Comment (seaona) [inline test/e2e/tests/hardware-wallets/lattice-connect.spec.ts:13]: nice correction

File: test/e2e/tests/hardware-wallets/trezor-sign.spec.ts [hunk 12]
```diff
@@ -26,28 +21,13 @@ describe('Trezor Hardware Signatures', function (this: Suite) {
...
-        );
-
-        assert.equal(
-          await verifyRecoverAddress.getText(),
-          KNOWN_PUBLIC_KEY_ADDRESSES[0].address.toLocaleLowerCase(),
+        await loginWithBalanceValidation(driver);
+        const testDappPage = new TestDappPage(driver);
+        await testDappPage.openTestDappPage();
+        await testDappPage.check_pageIsLoaded();
+        await testDappPage.signTypedDataV4(true);
```
Comment (seaona) [inline test/e2e/tests/hardware-wallets/trezor-sign.spec.ts:28]: I think passing an object with the flag name, would make it clearer what this boolean is actually indicating. However given that handling redesign/no-redesign is temporary I think it's fine as it is

**Review by seaona — APPROVED** [review]
LGTM!!

File: test/e2e/tests/hardware-wallets/trezor-sign.spec.ts (hunk 12 above)
Comment (chloeYue) [reply test/e2e/tests/hardware-wallets/trezor-sign.spec.ts:28]: yeah it's a very temporary solution, let's wait for the removal of the toggle as planned in the ticket https://github.com/MetaMask/MetaMask-planning/issues/3030

**Pull request merged**